pkg/upgrade/control_plane.go

```diff
@@ -673,18 +638,159 @@ func (u *ControlPlaneUpgrader) shouldSkipMachine(log logr.Logger, machine *clust
 	return false
 }
 
+// isHealthy verifies if a machine is healthy.
+// Conditions for health:
+// - the machine should have a ProviderID
+// - the machine should have a node associated with the ProviderID
+// - the node should be Ready
+// - the node should have the LabelNodeRoleMaster
+func (u *ControlPlaneUpgrader) isHealthy(machine *clusterv1.Machine) error {
+	ctx := context.TODO()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	newProviderID, err := u.waitForMachineWithProviderID(ctx, machine.Namespace, machine.Name)
+	if err != nil {
+		return err
+	}
+	node, err := u.waitForNodeWithProviderID(ctx, newProviderID)
+	if err != nil {
+		return err
+	}
+	if err := u.waitForNodeReady(ctx, node); err != nil {
+		return err
+	}
+
+	if err := u.waitForNodeWithMasterLabel(ctx, node); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// removeMachine removes the machine and its references.
+// It deletes and waits for the machine to be deleted.
+// If the machine has a ProviderID, it proceeds to remove the corresponding
+// etcd member and entry from the KubeadmConfigMap.
+func (u *ControlPlaneUpgrader) removeMachine(machine *clusterv1.Machine) error {
+	ctx := context.TODO()
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
+
+	log := u.log.WithValues(
+		"machine", fmt.Sprintf("%s/%s", machine.Namespace, machine.Name),
+		"upgrade", u.upgradeID,
+	)
+	const deleteMachineInterval = 10 * time.Second
+	var err error
+
+	log.Info("Deleting machine")
+	err = wait.PollImmediate(deleteMachineInterval, u.machineTimeout, func() (bool, error) {
```

Comment:
I think we can switch this to `wait.PollImmediateUntil`, so we can use the context created above instead of possibly blocking for 15 minutes.